### DISCUSSION TITLE
Support for Kotlin Multiplatform

### DISCUSSION THREAD

**Discussion by mcpierce** [readium/kotlin-toolkit#547] [Ideas]
I'm hoping to find that this is already happening, but has any thought been given to making the Kotlin library multiplatform compatible? I have a project that's targeting both iOS and Android and its core function is to work with a server that provides an OPDS API to access its content, and a cross-platform Readium library would be ideal for this.

**Comment by qnga** [comment]
Hi,
Yes, we want to do it. I doubt it will be completed before one year though, at least if Mickaël and I have to do all the work. Subliminal message: feel free to do it. We can provide you with guidance.

**Reply by mcpierce** [reply]
I would definitely like to participate in it, but don't think it's something I could tackle solo. TBH, I'm fairly new to mobile development (this app is my first big project of the type). I am a well-seasoned developer (30 years professional development) though and I learn pretty quickly.

**Reply by mcpierce** [reply]
In the meantime, I'd greatly appreciate a pointer to a good straight forward use case/example for v2.4.1 that I can use as a guide for my Android app. It look like the repositories are focused on v3 which I can't currently use.

**Reply by qnga** [reply]
Look at https://github.com/readium/kotlin-toolkit/discussions/54 (a bit outdated for 2.4 I guess) and https://github.com/readium/kotlin-toolkit/blob/2.4.1/test-app/src/main/java/org/readium/r2/testapp/reader/ReaderRepository.kt

**Reply by mcpierce** [reply]
How would I process a byte array if I've already downloaded it from the server? Basically, I've got the code in place to make the HTTP request to the server and get the content, and what I'm looking for is something to parse out the link details for the OPDS feed for me. Looking at the above I see how it seems to want a FileAsset (assuming by the name that's refering to a local file). Is there an Asset that can take a byte array as a stream?

**Reply by qnga** [reply]
The main use case of Readium is the opening and rendering of publications, that's the reason why I pointed you to the `ReaderRepository`. If you're more interested in OPDS, you can have a look at the [OPDS parsers](https://github.com/readium/kotlin-toolkit/tree/2.4.1/readium/opds/src/main/java/org/readium/r2/opds). I'm not aware of any documentation on that API.

Once you've downloaded a publication, you might want to store it in a local file. If you don't, you won't be able to open ZIP-based publications with v2.4, it's a v3 feature.

**Reply by mcpierce** [reply]
@qnga Thanks for that tip, it'll make working with the publications even easier! I didn't realize I'd be able to use Readium to interact with the content after retrieval and thought I'd only use it for online navigation.

Speaking of which, one last question: is there a way to add basic authentication to DefaultHttpClient? I think that might get me over this current hurdle if I can manage that.

**Reply by qnga** [reply]
Sure, you can pass a callback to the `DefaultHttpClient`. Change the request header in `onStartRequest`.

**Reply by mcpierce** [reply]
@qnga I'm hitting another issue where the parser is complaining about the server content. The server is returning OPDS XML content but, for some reason, the instance of OPDS2Parser is trying to treat it as JSON:

```
Caused by: org.json.JSONException: Value <?xml of type java.lang.String cannot be converted to JSONObject
```

**Reply by qnga** [reply]
Try the OPDS1Parser.  OPDS 1 is XML-based and OPDS2 is JSON-based.

**Reply by mcpierce** [reply]
That was the trick! Thank you, I'm now up and running with reading the OPDS stream from my server.

That being said, I've hit a bug with feed entries that include a URL encoded ":" and opened bug #548 to get it looked at. But otherwise I'm doing great now! Thank you so much for you help!

**Reply by mickael-menu** [reply]
> It look like the repositories are focused on v3 which I can't currently use.

If you're concerned by the beta status, I think by the time you release your app it will be stable. The current Kotlin v3 is pretty stable, we're just waiting for the Swift version to catch up before doing the proper release.

**Reply by mcpierce** [reply]
@mickael-menu Okay, that's good to know.

**Comment by RetRo99** [comment]
Is there any update about this?

**Reply by qnga** [reply]
Nothing has been done. On the other hand, I've been developing a Kotlin/Compose multiplatform app using the Kotlin toolkit on Android and the Swift one on iOS. That's tricky but it works.

**Reply by RetRo99** [reply]
Is the code available somewhere? I been looking into making a client for [StoryTeller](https://storyteller-platform.gitlab.io/storyteller/) in KMP. I would be willing to help, if help is needed for KMP support.

**Reply by qnga** [reply]
No, it is not. I'd like to publish the multiplatform wrapper but I don't have time to extract it at the moment. Do you plan to use Compose Multiplatform or Swift UI for the iOS UI? The best approach may depend on it.

**Reply by RetRo99** [reply]
I wanted to use Compose Multiplatform for everything but the EPUB rendering.

**Reply by qnga** [reply]
Ok, so you're in a situation similar to mine. Our first goal in making the toolkit multiplatform concerns only the shared models and the parsers, so even if it was done, you'd still have to use the Swift navigator (as you spotted it) with Compose, which is, I think, the trickiest part. So if I were you, I'd just take the toolkit as it is, as I did.

Concretely, I'm using a trick suggested by Mickaël: thanks to dependency inversion, I don't use Swift API in Kotlin at all, only Kotlin API in Swift. Every Swift component I need is wrapped into a new Swift component implementing a Kotlin interface. At startup, the implementations of these interfaces are passed from the Swift iOS entry point to the Kotlin/Compose code which can use them when they are needed. This approach works well but there is a bunch of boilerplate. It is not entirely clear if this is the simplest approach.

**Reply by mickael-menu** [reply]
I'm also working on a KMP app, but I use a pure SwiftUI implementation for the iOS UI. This makes things easier since I can use Swift Navigators directly. However, I wrote wrappers for everything else, including the parsers, asset retriever, and `Publication` models. We don't plan to publish the wrapper for now:

- I think this is a temporary solution. I'd prefer to use the Kotlin toolkit directly in my Kotlin layer and simply use an adapter for the Swift navigators.
- I only wrapped the necessary elements. For example I limited the `Publication` models to what I needed for our bookshelf import.

> I would be willing to help, if help is needed for KMP support.

Unfortunately, we don't have much time to adapt the Kotlin toolkit for KMP support. We welcome contributions to help us with this :) Ideally, we should use small, incremental PRs to progressively remove JVM and Android dependencies from the modules. This approach allows us to release KMP-ready modules one by one as they are completed. The first step will likely involve adapting the Publication models (https://github.com/readium/kotlin-toolkit/tree/develop/readium/shared/src/main/java/org/readium/r2/shared/publication). Right now they are in the `readium-shared` modules, but I think it would make sense to create a dedicated module for the Readium Web Publication Models.

**Reply by yuroyami** [reply]
> At startup, the implementations of these interfaces are passed from the Swift iOS entry point to the Kotlin/Compose code which can use them when they are needed. This approach works well but there is a bunch of boilerplate. It is not entirely clear if this is the simplest approach.


I'm running into the same boilerplate issue in my own KMP app. Much of it could be eliminated if parts of Readium were commonized. For instance, `swift-toolkit` uses SwiftSoup, which has a KMP equivalent in KSoup — and that's just one example among many. 

As for rendering itself, is there a chance Compose be leveraged? The toolkits are also tightly coupled to platform-specific view systems (Fragments, etc.). Rendering through a `ComposeView` instead would make integration straightforward in both Fragments and Activities. But that's just speaking in Android language. 

Commonizing the portable parts of Readium would meaningfully modernize the library. KMP may still be niche, but it's worth noting that Kotlin libraries are already consumable from Swift — so commonizing in Kotlin wouldn't lock out iOS-only or Android-only projects. It would also simplify your own maintenance, since shared logic only needs to be updated in one place.

**Reply by qnga** [reply]
Though I think the Swift toolkit will always be useful for apps targeting only iOS or using native UI, at least for the rendering parts (it's more straightforward), I'm 100% convinced of the benefits of a Kotlin multiplatform toolkit. I wrote a new rendering engine using Compose, but it's Android only at the moment, as the rest of the toolkit. I'd be delighted to make it multiplatform (iOS and desktop) if you know someone interested in sponsoring that (quite big) work.

**Reply by mickael-menu** [reply]
@yuroyami You don't need to convince us 😄 We (both maintainers) of the mobile toolkits are already using Readium in two independent KMP apps. There's just a lot of things to tackle for our bandwidth. Personally I'm focused on iOS this year, and should pivot to Android by the end of 2026. By then, I'd like to experiment with converting shared models to KMP.

As @qnga mentioned, we're also open to the idea of building a Compose Multiplatform navigator in the Kotlin toolkit. It's not a priority though. This could be a sponsoring or a third-party contribution, once the rest of the toolkit is KMP-ready.

That being said, it's unlikely that we will replace the "backend" components of the Swift toolkit with KMP. At this point most of the Swift shared models and utilities are in maintenance mode, and the grunt of the work is on improving the navigators. However, I'd like for the Swift navigators to be eventually independent from the source of the `Publication` (either from the Swift shared models, or the KMP toolkit).

**Reply by yuroyami** [reply]
@qnga I'll definitely reach out if a sponsorship opportunity comes up, but honestly I suspect I'll end up being the one jumping in with you regardless, I genuinely enjoy writing KMP and Readium has already saved me a lot of trouble, so contributing back in a meaningful way feels like the right thing to do. Commonizing it would be something I'd take real pleasure in working on.

@mickael-menu I think the "write once, use everywhere" angle is worth dwelling on from a maintainer perspective specifically: shared logic lives in one place, gets updated once and propagates everywhere, and gets exported as a KLib that both Swift/iOS and Kotlin/Android consume without friction. From a consumer standpoint nothing really changes — it's still a native library with Objective-C interop performance, no overhead, fully consumable from Swift, and distributable as an XCFramework via CocoaPods or SPM. The only thing that changes for you is that the source of truth is clean Kotlin in a single codebase rather than two diverging implementations. I may be biased as someone who prefers Kotlin over just about everything else, but I genuinely think it's worth serious consideration.

Either way, I'm hoping to see this happen someday and looking forward to the point where I can contribute the way I'd like to. Best of luck to both of you in the meantime! 🙂
